pkg/expect/expect.go

```diff
@@ -187,9 +193,26 @@ func (ep *ExpectProcess) ExpectFunc(ctx context.Context, f func(string) bool) (s
 		}
 	}
 
+	select {
+	// NOTE: The ep.read() goroutine needs the mutex so we have to drain the
+	// log without lock.
+	case <-ep.readCloseCh:
+	case <-ctx.Done():
+		return "", fmt.Errorf("failed to find match string: %w", ctx.Err())
+	}
+
 	ep.mu.Lock()
 	defer ep.mu.Unlock()
 
+	// retry it since we get all the log data
+	for i < len(ep.lines) {
+		line := ep.lines[i]
+		i++
+		if f(line) {
+			return line, nil
```

Comment:
I don't think any consumer of `ExpectFunc` relies on the exit error, so return nil should be fine.